### PR TITLE
Add some stuff for AIs

### PR DESCRIPTION
**Pull request by liach**

### CODE REVIEW

**Comment by asiekierka** [comment]
Since you're a new contributor - can you confirm you haven't taken the mappings from external sources like MCP?

**Comment by liach** [comment]
How to confirm that? Those tasks are named `EntityAIXx` in mcp, but I named them `Goal`s. Maybe `AiTask` are also good names, but as I see vanilla seems to name them `Goal`. And I've read that google doc in the readme.

File: mappings/net/minecraft/entity/ai/IDistantAttacker.mapping [hunk 1]
```diff
@@ -0,0 +1,3 @@
+CLASS none/zq net/minecraft/entity/ai/IDistantAttacker
```
Comment (liach) [inline mappings/net/minecraft/entity/ai/IDistantAttacker.mapping:1]: Well I just found that I need to change this to `DistantAttacker`

File: mappings/none/zb.mapping [hunk 2]
```diff
@@ -1,5 +0,0 @@
-CLASS none/zb
```
Comment (liach) [inline mappings/none/zb.mapping]: Well this becomes `EntitySkeletonBase` but git showed this as deleted instead of renamed.

**Comment by modmuss50** [comment]
As long as you haven't looked at MCP and have had a think about what the best name would be without refrancing MCP you should be ok. Sometimes names will be the same, but as long as you have thought up the name it will be ok.

At the end of the day your the only person who can say if you have used MCP or not.

And feel free to ask on IRC if you have ANY questions.

File: mappings/none/zb.mapping (hunk 2 above)
Comment (modmuss50) [reply mappings/none/zb.mapping]: Git can be weird about this type of thing some times.

**Comment by asiekierka** [comment]
@liach I cannot read through the PR until you say yes.

Have you referenced MCP (as in the files) while working on the mappings?

**Comment by liach** [comment]
I did not refer to MCP while I work.

File: mappings/net/minecraft/entity/ai/GoalSet.mapping [hunk 3]
```diff
@@ -0,0 +1,41 @@
+CLASS none/tz net/minecraft/entity/ai/GoalSet
+	CLASS none/tz$a WeighedGoal
```
Comment (asiekierka) [inline mappings/net/minecraft/entity/ai/GoalSet.mapping]: WeightedGoal?

File: mappings/net/minecraft/entity/ai/GoalSet.mapping [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+	METHOD a tick ()V
+	METHOD a add (ILnone/ty;)V
+		ARG 0 weight
+		ARG 1 task
+	METHOD a changeBits (IZ)V
+		ARG 0 taskBits
+		ARG 1 remove
+	METHOD a remove (Lnone/ty;)V
+		ARG 0 task
+	METHOD a shouldContinueExecuting (Lnone/tz$a;)Z
```
Comment (asiekierka) [inline mappings/net/minecraft/entity/ai/GoalSet.mapping]: shouldContinueExecution?

File: mappings/net/minecraft/entity/ai/IDistantAttacker.mapping (hunk 1 above)
Comment (asiekierka) [reply mappings/net/minecraft/entity/ai/IDistantAttacker.mapping:1]: If it's an interface, no.

File: mappings/net/minecraft/entity/ai/GoalSet.mapping (hunk 3 above)
Comment (liach) [reply mappings/net/minecraft/entity/ai/GoalSet.mapping]: A goal with a priority integer, smaller ones seem to execute first.

**Comment by copygirl** [comment]
Mojang's mappings are just a loose guideline, so I suggest `AiGoal` to make the intention of the class clearer. Though I do like `AiTask` more because not every task is a "goal", like looking at another entity.

File: mappings/net/minecraft/entity/ai/GoalBase.mapping [hunk 5]
```diff
@@ -0,0 +1,11 @@
+CLASS none/ty net/minecraft/entity/ai/GoalBase
```
Comment (copygirl) [inline mappings/net/minecraft/entity/ai/GoalBase.mapping]: Get rid of the `Base`?

File: mappings/net/minecraft/entity/ai/impl/mob/AiTaskSkeletonHorse.mapping [hunk 6]
```diff
@@ -0,0 +1,4 @@
+CLASS none/xh net/minecraft/entity/ai/impl/mob/AiTaskSkeletonHorse
```
Comment (copygirl) [inline mappings/net/minecraft/entity/ai/impl/mob/AiTaskSkeletonHorse.mapping]: This is still named `AiTask`.

File: mappings/net/minecraft/entity/passive/EntityPetBase.mapping [hunk 7]
```diff
@@ -1,3 +1,3 @@
-CLASS none/sy
+CLASS none/sy net/minecraft/entity/passive/EntityPetBase
```
Comment (copygirl) [inline mappings/net/minecraft/entity/passive/EntityPetBase.mapping]: I've renamed this class to `EntityTameable` my PR already.

**Comment by asiekierka** [comment]
I agree with AiGoal.

File: mappings/net/minecraft/entity/ai/impl/mob/AiTaskSkeletonHorse.mapping (hunk 6 above)
Comment (liach) [reply mappings/net/minecraft/entity/ai/impl/mob/AiTaskSkeletonHorse.mapping]: Yeah some I did not rename necessarily.

**Comment by copygirl** [comment]
I'll ask everyone more directly again now that the change to `AiGoal` was done. In my opinion, `AiTask` represents the nature of these things better, as they include some "idle" tasks that aren't necessarily "goals" (from what I know of them, anyway, which isn't much). Like walking randomly, looking at things, ...

:+1: for `AiTask` and :-1: for `AiGoal` ?

**Comment by asiekierka** [comment]
@copygirl Looking at something is very much a goal the AI has to meet!

**Comment by copygirl** [comment]
@asiekierka Additionally, the methods on the class on the class remind me of a computing task (`shouldContinueExecution`) and it it tells the entity what to do rather than describing the goal itself.

**Comment by liach** [comment]
@copygirl Once I wanted to name them `AiTask`, as Sponge API has named them that way, but I wanted to use `Goal` because in code it mentioned Goals.

**Comment by asiekierka** [comment]
Can we get something done about this branch soonish? I really want to update to 16w36a, as it has a lot of rendering fixes and maybe refactors too!

File: mappings/net/minecraft/entity/boss/impl/EntityWither.mapping.orig [hunk 8]
```diff
@@ -0,0 +1,38 @@
+CLASS none/yj net/minecraft/entity/boss/impl/EntityWither
```
Comment (shadowfacts) [inline mappings/net/minecraft/entity/boss/impl/EntityWither.mapping.orig]: Remove this file

File: mappings/net/minecraft/entity/mob/EntitySkeletonBase.mapping.orig [hunk 9]
```diff
@@ -0,0 +1,21 @@
+CLASS none/zb net/minecraft/entity/mob/EntitySkeletonBase
```
Comment (shadowfacts) [inline mappings/net/minecraft/entity/mob/EntitySkeletonBase.mapping.orig]: Remove this file

**Comment by shadowfacts** [comment]
Mappings LGTM, but there are a whole bunch of .orig and backup files that need to be removed.

File: mappings/net/minecraft/entity/passive/impl/EntitySquid.mapping.orig [hunk 10]
```diff
@@ -0,0 +1,12 @@
+CLASS none/ww net/minecraft/entity/passive/impl/EntitySquid
+<<<<<<< f857ed83de0d5eb3e92f0b4927e58712e160c119
```
Comment (modmuss50) [inline mappings/net/minecraft/entity/passive/impl/EntitySquid.mapping.orig]: This is broken

File: mappings/net/minecraft/entity/mob/EntitySkeletonBase_BACKUP_10312.mapping [hunk 11]
```diff
@@ -0,0 +1,21 @@
+CLASS none/zb net/minecraft/entity/mob/EntitySkeletonBase
```
Comment (modmuss50) [inline mappings/net/minecraft/entity/mob/EntitySkeletonBase_BACKUP_10312.mapping]: Remove this as well

File: mappings/net/minecraft/entity/mob/EntitySkeletonBase_BACKUP_740.mapping [hunk 12]
```diff
@@ -0,0 +1,21 @@
+CLASS none/zb net/minecraft/entity/mob/EntitySkeletonBase
```
Comment (modmuss50) [inline mappings/net/minecraft/entity/mob/EntitySkeletonBase_BACKUP_740.mapping]: And this

File: mappings/net/minecraft/entity/mob/EntitySkeletonBase_LOCAL_10312.mapping [hunk 13]
```diff
@@ -0,0 +1,10 @@
+CLASS none/zb net/minecraft/entity/mob/EntitySkeletonBase
```
Comment (modmuss50) [inline mappings/net/minecraft/entity/mob/EntitySkeletonBase_LOCAL_10312.mapping]: Doesnt seem like the correct file name

**Comment by modmuss50** [comment]
I would  go though all the files, but there are a load that should not be there

File: mappings/net/minecraft/entity/passive/impl/EntitySquid.mapping.orig (hunk 10 above)
Comment (liach) [reply mappings/net/minecraft/entity/passive/impl/EntitySquid.mapping.orig]: `EntitySquid.mapping.orig` this is not a mapping but a ref for the broken one.

**Comment by liach** [comment]
Actually those files are references generated by merge tools.

**Comment by liach** [comment]
So do I need to apply @copygirl 's suggestion right now?

**Comment by copygirl** [comment]
@liach You can keep it the way it is for now. It's always possible to change in the future. I'll probably put in on my [personal todo list](https://github.com/copygirl/pomf/issues), make a PR, and move the discussion there.

**Comment by liach** [comment]
Thanks.

File: mappings/net/minecraft/entity/boss/impl/EntityWither.mapping [hunk 14]
```diff
@@ -9,12 +9,13 @@ CLASS none/yj net/minecraft/entity/boss/impl/EntityWither
 		ARG 0 headIndex
 		ARG 1 value
 	METHOD a canDestroy (Lnone/akw;)Z
-		ARG 0 block
-	METHOD a deserializeEntityTag (Lnone/dt;)V
+		ARG 0 block	METHOD a deserializeEntityTag (Lnone/dt;)V
```
Comment (asiekierka) [inline mappings/net/minecraft/entity/boss/impl/EntityWither.mapping]: oh no you've merged it wrong

File: mappings/net/minecraft/entity/boss/impl/EntityWither.mapping [hunk 15]
```diff
@@ -1,5 +1,5 @@
 CLASS none/yj net/minecraft/entity/boss/impl/EntityWither
-	FIELD a TRACKED_ENTITY_ID_1 Lnone/kj;
+	CLASS none/yj$a		METHOD a canStart ()Z	FIELD a TRACKED_ENTITY_ID_1 Lnone/kj;
```
Comment (asiekierka) [inline mappings/net/minecraft/entity/boss/impl/EntityWither.mapping]: oh no you've merged it wrong

File: mappings/net/minecraft/entity/mob/EntitySkeletonBase_REMOTE_10312.mapping [hunk 16]
```diff
@@ -0,0 +1,10 @@
+CLASS none/zb net/minecraft/entity/mob/EntitySkeletonBase
```
Comment (asiekierka) [inline mappings/net/minecraft/entity/mob/EntitySkeletonBase_REMOTE_10312.mapping]: missed a file ALSO all the _BASE_ files

File: mappings/net/minecraft/entity/mob/EntitySkeletonBase_REMOTE_740.mapping [hunk 17]
```diff
@@ -0,0 +1,10 @@
+CLASS none/zb net/minecraft/entity/mob/EntitySkeletonBase
```
Comment (asiekierka) [inline mappings/net/minecraft/entity/mob/EntitySkeletonBase_REMOTE_740.mapping]: missed a file

File: mappings/net/minecraft/entity/mob/impl/EntityEnderman.mapping [hunk 18]
```diff
@@ -1,12 +1,28 @@
...
+		METHOD e tick ()V
 	FIELD bA ageWhenTargetSet I
 	FIELD bx CARRIED_BLOCK Lnone/kj;
 	FIELD by ANGRY Lnone/kj;
 	FIELD c CARRIABLE_BLOCKS Ljava/util/Set;
 	METHOD G getSoundAmbient ()Lnone/nk;
 	METHOD a setCarriedBlock (Lnone/asj;)V
-		ARG 0 value
-	METHOD a deserializeEntityTag (Lnone/dt;)V
+		ARG 0 value	METHOD a deserializeEntityTag (Lnone/dt;)V
```
Comment (asiekierka) [inline mappings/net/minecraft/entity/mob/impl/EntityEnderman.mapping]: meeeerged it wrong

File: mappings/net/minecraft/entity/mob/impl/EntityGhast.mapping [hunk 19]
```diff
@@ -1,6 +1,18 @@
...
+		METHOD c start ()V
+		METHOD d onRemove ()V
+		METHOD e tick ()V
+	CLASS none/zj$d
+		METHOD a canStart ()Z
+		METHOD b shouldContinue ()Z
+		METHOD c start ()V
 	FIELD a SHOOTING Lnone/kj;
-	METHOD G getSoundAmbient ()Lnone/nk;
+	METHOD G getSoundDeath ()Lnone/nk;
```
Comment (asiekierka) [inline mappings/net/minecraft/entity/mob/impl/EntityGhast.mapping]: merged it wrong again, this would make minecraft crash

**Comment by asiekierka** [comment]
My personal recommendation for merging is to copy the changed files, enforce the "theirs" strategy, and re-introduce the changes manually with Enigma again.

**Comment by liach** [comment]
Well merging those files are a problem. Will fix.

**Comment by asiekierka** [comment]
Let's hope.

**Comment by asiekierka** [comment]
But please, next time, at least try to open the thing in Enigma.

**Comment by liach** [comment]
Okay.

**Pull request merged**